src/Workspaces/Core/Portable/Execution/AbstractOptionsSerializationService.cs

```diff
@@ -204,5 +210,112 @@ protected void WriteParseOptionsTo(ParseOptions options, ObjectWriter writer, Ca
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var value = options.GetOption(option);
+            writer.WriteString(value.ToXElement().ToString());
+        }
+
+        protected OptionSet ReadOptionFrom<T>(OptionSet options, Option<CodeStyleOption<T>> option, ObjectReader reader, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var xmlText = reader.ReadString();
+            var value = CodeStyleOption<T>.FromXElement(XElement.Parse(xmlText));
+
+            return options.WithChangedOption(option, value);
+        }
+
+        private void WriteOptionTo<T>(OptionSet options, string language, PerLanguageOption<CodeStyleOption<T>> option, ObjectWriter writer, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var value = options.GetOption(option, language);
+            writer.WriteString(value.ToXElement().ToString());
+        }
+
+        private OptionSet ReadOptionFrom<T>(OptionSet options, string language, PerLanguageOption<CodeStyleOption<T>> option, ObjectReader reader, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var xmlText = reader.ReadString();
+            var value = CodeStyleOption<T>.FromXElement(XElement.Parse(xmlText));
+
+            return options.WithChangedOption(option, language, value);
+        }
+
+        /// <summary>
+        /// this is not real option set. it doesn't have all options defined in host. but only those
+        /// we pre-selected.
+        /// </summary>
+        protected class SerializedPartialOptionSet : OptionSet
+        {
+            private readonly ImmutableDictionary<OptionKey, object> _values;
+
+            public SerializedPartialOptionSet()
+            {
+                _values = ImmutableDictionary<OptionKey, object>.Empty;
+            }
+
+            private SerializedPartialOptionSet(ImmutableDictionary<OptionKey, object> values)
+            {
+                _values = values;
+            }
+
+            public override object GetOption(OptionKey optionKey)
+            {
+                object value;
+                Contract.ThrowIfFalse(_values.TryGetValue(optionKey, out value));
+
+                return value;
+            }
+
+            public override OptionSet WithChangedOption(OptionKey optionAndLanguage, object value)
+            {
+                return new SerializedPartialOptionSet(_values.SetItem(optionAndLanguage, value));
+            }
+
+            internal override IEnumerable<OptionKey> GetChangedOptions(OptionSet optionSet)
+            {
+                foreach (var kvp in _values)
+                {
+                    var currentValue = optionSet.GetOption(kvp.Key);
```

Comment:
I think we should enforce T of Option<T> or PerLanguageOptoin<T> to be either IEquatable<T> or enum. otherwise, GetChangedOption won't work.